src/common/fault_domain.h

```diff
@@ -0,0 +1,128 @@
+/**
+ * (C) Copyright 2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#ifndef __DAOS_COMMON_FAULT_DOMAIN__
+#define __DAOS_COMMON_FAULT_DOMAIN__
+
+#include <daos/common.h>
+
+/**
+ * Possible types of fault domain tree node.
+ */
+enum d_fd_node_type {
+	D_FD_NODE_TYPE_UNKNOWN = 0,
+	D_FD_NODE_TYPE_DOMAIN,
+	D_FD_NODE_TYPE_RANK,
+};
+
+/**
+ * Unified structure representing a node in the fault domain tree.
+ *
+ */
+struct d_fd_node {
+	enum d_fd_node_type fdn_type; /* indicates which union member to use */
+	union {
+		const struct d_fault_domain		*dom;
+		const struct d_fault_domain_rank	*rank;
+	} fdn_val;
+};
+
+/**
+ * Represents a fault domain at any level above rank.
+ */
+struct d_fault_domain {
+	uint32_t fd_level;	/** level in the fault domain tree >= 1 */
+	uint32_t fd_id;		/** unique ID */
+	uint32_t fd_children_nr; /** number of children */
+};
+
+/** Represents a fault domain rank */
+struct d_fault_domain_rank {
+	uint32_t fdr_rank; /** rank number */
+};
+
+/**
+ * DAOS fault domain tree. This is a wrapper for the compressed integer array
+ * format passed down from the control plane.
+ * See src/proto/mgmt/pool.proto for details.
+ */
+struct d_fd_tree {
```

Comment:
I tried this, but as I was implementing, I thought it was less trouble to let callers allocate the structure on the stack rather than the heap if desired. Opaque structs only seem to work with pointers. I can add a comment that the internals shouldn't be used as an interface.